update lecture&course of course7 (#1)
moonbitlang/moonbit-course: course7/course_en.md

> # Modern Programming Thoughts: Imperative Programming

Reviewer: Modern Programming Ideology would be a better name.

```

## Commands and Side Effects

The `print` command in MoonBit enables us to output strings, such as `print("hello moonbit")`. We can define initialization instructions in the `init` code block, serving as the program entry point. However, output commands may break referential transparency.

```moonbit
fn init {
  let x: Int = {
    println("hello moonbit") // First, execute the command and perform output
    1 + 1 // Then, take the last value of the expression block as the block's value
  }
  let z: Int = square(x) // Output once
}
```

We typically use the unit type `Unit` to represent the return value of functions or commands with side effects. It has only one value: `()`.

## Variables and Aliases

In MoonBit, we can define temporary variables using `let mut`. The assignment operation itself is a command. Struct fields are immutable by default, but we can use `mut` to identify mutable fields, which are treated as references.

```moonbit
struct Ref[T] { mut val : T }

fn init {
  let ref: Ref[Int] = { val: 1 } // ref itself is just a data binding
  ref.val = 10 // We can modify the fields of the struct
  println(ref.val.to_string()) // Output 10
}
```

> Multiple identifiers pointing to the same mutable data structure can be considered aliases, which need to be handled carefully.

Reviewer: Examples are missing and the explanation is too simple.

## Loops

We can define loops using variables in MoonBit. A loop includes defining the loop variable and its initial value, checking whether to continue the loop, and iterating the variable.

```moonbit
let mut i = 0

> while i < 2, i = i + 1 {

Reviewer: The grammar of while loop has been changed lately. Now it should be
```rust
let mut i = 0
while i < 2 {
  println("Output")
  i = i + 1
}
```

  println("Output")
} // Repeat output 2 times
```

> The loop execution flow is: check the condition -> execute commands -> iterate the variable -> repeat the above process. MoonBit provides a debugger that enables observing the changes in runtime data during the loop process.

Reviewer: How to use a debugger?

## Loops and Recursion

Loops and recursion are equivalent. A loop can be written in a recursive form. Taking the calculation of the Fibonacci sequence as an example:

```moonbit
// Loop form
fn fib_mut(n: Int) -> Int {

>   let mut a = 0; let mut b = 1; let mut i = 0

Reviewer: Since this is a document not a presentation, you may write them in multiple lines.

  while i < n, i = i + 1 {
    let t = a + b
    a = b; b = t
  }
  a
}

// Recursive form
fn fib(n: Int) -> Int {
  if n == 0 { 0 }
  else if n == 1 { 1 }
  else { fib(n - 1) + fib(n - 2) }
}
```

## Controlling Loop Flow

Within a loop, we can use `break` to exit the loop prematurely and `continue` to skip the remaining part of the current loop iteration.

```moonbit
fn print_first_3() {
  let mut i = 0
  while i < 10, i = i + 1 {
    if i == 3 {
      break // Skip from 3 onwards
    } else {
      println(i.to_string())
    }
  }
}

> ```

Reviewer: Explanation of continue?
Also is the expected output of the snippet?

> ## MoonBit Checks

Reviewer: Weird title

# Modern Programming Ideology: Imperative Programming

> 

Reviewer: Maybe a gentle overall introduction here?

Author: but the next part is

Imperative programming is a style of programming where we give the computer a set of commands to execute, much like giving instructions to a person. It's about telling the computer exactly what to do, step by step. This is different from functional programming, where you define what you want in terms of mathematical functions. In functional programming, the same input will always give you the same output, and there's a concept called "referential transparency," which means you can replace a function call with its result without changing the program's behavior.

However, in the real world, we often need our programs to do more than just calculations. We might need them to read from a file, write to the screen, or change data in memory. These actions are called "side effects," and they can make our programs behave differently each time they run, even with the same input. This breaks the referential transparency, making our programs harder to understand and predict.

```moonbit
let x: Int = 1 + 1 // x can be directly replaced with 2
fn square(x: Int) -> Int { x * x }
let z: Int = square(x) // can be replaced with { 2 * 2 }, still resulting in 4
```

> Let's look at an example in MoonBit, a fictional programming language. We can declare a variable `x` and assign it the result of an expression, like `1 + 1`. This is straightforward and mirrors what we do in mathematics. We can also define a function `square` that takes an integer and returns its square. When we call `square(x)`, we get the result `4`, just like if we replaced `square(x)` with `{ 2 * 2 }`.

Reviewer: ```suggestion
Let's look at the example here in MoonBit. We can declare a variable `x` and assign it the result of an expression, like `1 + 1`. This is straightforward and mirrors what we do in mathematics. We can also define a function `square` that takes an integer and returns its square. When we call `square(x)`, we get the result `4`, just like if we replaced `square(x)` with `{ 2 * 2 }`.
```

## Commands and Side Effects

In MoonBit, we can use the `print` command to output text to the screen. This is a side effect because it changes the state of the world outside our program. When we define a function, we usually want it to return a value, but sometimes we just want to perform an action without a return value. In such cases, we use the `Unit` type, which represents the absence of a meaningful return value, like a placeholder.

```moonbit
fn init {
  let x: Int = {
    println("hello moonbit") // First, execute the command and perform output
    1 + 1 // Then, take the last value of the expression block as the block's value
  }
  let z: Int = square(x) // Output once
}
```

> Here's a simple example: we define a function `init` that prints a greeting message and then calculates the value of `x`. The `println` command is executed first, and then the result of the expression `1 + 1` is used as the value of `x`. This function has a side effect (printing), but it also returns a value (the value of `x`), which is why we use the `Unit` type to indicate that the primary purpose of this function is not to return a value.

Reviewer: > This function has a side effect (printing), but it also returns a value (the value of `x`), which is why we use the `Unit` type to indicate that the primary purpose of this function is not to return a value.

Doesn't make sense. This function `init` does not return any value. Also `x` is defined with a code block, if that's what you mean.

> Here's a simple example: we define a function `init` that prints a greeting message and then calculates the value of `x`. The `println` command is executed first, and then the result of the expression `1 + 1` is used as the value of `x`. This function has a side effect (printing), but it also returns a value (the value of `x`), which is why we use the `Unit` type to indicate that the primary purpose of this function is not to return a value.

Reviewer: ```suggestion
Here's a simple example: we define a function `init`, a function that will be executed at the initialization stage of the program, that prints a greeting message and then calculates the value of `x`. The `println` command is executed first, and then the result of the expression `1 + 1` is used as the value of `x`. This function has a side effect (printing), but it also returns a value (the value of `x`), which is why we use the `Unit` type to indicate that the primary purpose of this function is not to return a value.
```

## Variables and Aliases

> In MoonBit, we can create temporary variables using `let mut`. The `mut` keyword tells MoonBit that the variable can change. This is important because it affects how we think about the data. For example, if we have a struct (a collection of values) called `Ref`, we can create a mutable reference to an integer inside it. We can then change the value of that integer, and any other references to the same `Ref` struct will see the updated value. This is because they are aliases, or different names for the same data.

Reviewer: ```suggestion
In MoonBit, we can create mutable variables using `let mut`. The `mut` keyword tells MoonBit that the variable can change. This is important because it affects how we think about the data. For example, if we have a struct called `Ref`, we can create a mutable reference with an integer inside it. We can then change the value of that integer, and any other references to the same `Ref` struct will see the updated value. This is because they are aliases, or different names for the same data.
```

```moonbit
struct Ref[T] { mut val : T }

fn init {
  let ref: Ref[Int] = { val: 1 } // ref itself is just a data binding
  ref.val = 10 // We can modify the fields of the struct
  println(ref.val.to_string()) // Output 10
}
```

Multiple identifiers pointing to the same mutable data structure can be considered aliases, which need to be handled carefully.

```moonbit
fn alter(a: Ref[Int], b: Ref[Int]) {
  a.val = 10
  b.val = 20
}

> 

Reviewer: illustration missing

fn init {
  let x: Ref[Int] = { val : 1 }
  alter(x, x)
  println(x.val.to_string()) // x.val will be changed twice
}
```

## Debugger

Debugging is like being a detective. You're trying to figure out why your program isn't working as expected. MoonBit's debugger is a tool that helps you do this by showing you what's happening inside your program as it runs. You can pause the program at any point, look at the values of variables, and step through the code one line at a time. This is incredibly useful for understanding complex behavior and fixing bugs.

![](../pics/debugger.png)

## Loops

> Loops are a way to repeat a block of code multiple times. In MoonBit, we can use a `while` loop to do this. We start by defining a loop variable, say `i`, and giving it an initial value. Then we check a condition—if the condition is true, we execute the commands inside the loop, and then we update the loop variable. This process repeats until the condition is no longer true, at which point the loop ends.

Reviewer: ```suggestion
Loops are a way to repeat a block of code multiple times. In MoonBit, we can use a `while` loop to do this. We start by defining a loop variable, say `i`, and giving it an initial value. Then we check the condition defined in the `while` loop. If the condition is true, we execute the commands inside the `while` loop's body. This process repeats until the condition is no longer true, at which point the loop ends. To avoid an infinite loop, we need to update the loop variable inside the loop body.
```

```moonbit
let mut i = 0
while i < 2 {
  println("Output")
  i += 1
} // Repeat output 2 times
```

For example, we might want to print the numbers from `0` to `1`. We can do this with a loop that checks if `i` is less than `2`. If it is, we print the current value of `i`, increment `i` by `1`, and then repeat the process. The loop continues until `i` is no longer less than `2`, and then we're done.

> ## Loops and Recursion

Reviewer: ```suggestion
### Loops and Recursion
```

> Loops and recursion are equivalent. A loop is a set of instructions that we repeat until a certain condition is met, while recursion is a way of solving a problem by breaking it down into smaller, similar problems. In programming, we can often rewrite a loop as a recursive function and vice versa.

Reviewer: ```suggestion
Loops and recursions are equivalent. A loop is a set of instructions that we repeat until a certain condition is met, while recursion is a way of solving a problem by breaking it down into smaller, similar problems. In programming, we can often rewrite a loop as a recursive function and vice versa.
```

For example, let's consider calculating the Fibonacci sequence. We can do this with a loop that keeps track of the last two numbers and updates them as it goes. Alternatively, we can write a recursive function that calls itself with smaller numbers until it reaches the base cases (0 and 1).

```moonbit
// Loop form
fn fib_mut(n: Int) -> Int {
  let mut a = 0;
  let mut b = 1;
  let mut i = 0
  while i < n {
    let t = a + b
    a = b; b = t
    i += 1
  }
  a
}

> // Recursive form

Reviewer: Not equivalent. If you don't know how to write an equivalent version, just use the same example as the presentation.

Author: the output of these two functinos is equivalent

Reviewer: The purpose of this section is to describe how to write the loop with a tail recursion function.

Author: > The purpose of this section is to describe how to write the loop with a tail recursion function.

but why put a Fibonacci function as example here inside chinese version

Reviewer: I didn't put a Fibonacci function here, that's also why I said you may use the example as the presentation. Also, since you've mentioned, you forgot to put the fibonacci example in the 'Mutable data' section (which was to demonstrate that having mutable data structure can also lead to referential transparency)

fn fib(n: Int) -> Int {
  if n == 0 { 0 }
  else if n == 1 { 1 }
  else { fib(n - 1) + fib(n - 2) }
}
```

> ## Controlling Loop Flow

Reviewer: ```suggestion
### Controlling Loop Flow
```

Sometimes we want to control the flow of a loop more precisely. We might want to skip the rest of the current iteration or exit the loop entirely. In MoonBit, we can use `break` to exit a loop early or `continue` to skip the rest of the current iteration and move on to the next one.

For example, if we're printing numbers from `0` to `9`, but we want to skip the number `3`, we can use `continue` in the loop condition. If we want to stop the loop entirely when we reach `3`, we can use `break`.

```moonbit
fn print_first_3() {
  let mut i = 0
  while i < 10 {
    i += 1
    print(i)
    println(" yes")
    if i == 3 {
      break // Skip from 3 onwards
    } else {
      println(i.to_string())
    }
  }
}
```

the excepted output is

```
1 yes
2 yes
```

but if we change `break` to `continue`

```moonbit
fn print_first_3() {
  let mut i = 0
  while i < 10 {
    i += 1
    print(i)
    println(" yes")
    if i == 3 {
      continue // go into the next iteration
    } else {
      println(i.to_string())
    }
  }
}
```

the excepted output is

```
1 yes
2 yes
 yes
 yes
 yes
 yes
 yes
 yes
 yes
```

> ## MoonBit Check

Reviewer: ```suggestion
```

MoonBit has some built-in checks to help us write better code. It checks if a variable that's supposed to be mutable has been modified, which can help us catch mistakes like forgetting to update a loop counter. It also checks that the return value of a function matches the declared return type, which helps us avoid type errors.

> 

Reviewer: screenshot here

```moonbit
let mut i = 0
while i < 2 {
  println("Output")
  i += 1
} // Repeat output 2 times
```

For example, we might want to print the numbers from `0` to `1`. We can do this with a loop that checks if `i` is less than `2`. If it is, we print the current value of `i`, increment `i` by `1`, and then repeat the process. The loop continues until `i` is no longer less than `2`, and then we're done.

### Loops and Recursion

Loops and recursions are equivalent. A loop is a set of instructions that we repeat until a certain condition is met, while recursion is a way of solving a problem by breaking it down into smaller, similar problems. In programming, we can often rewrite a loop as a recursive function and vice versa.

For example, let's consider calculating the Fibonacci sequence. We can do this with a loop that keeps track of the last two numbers and updates them as it goes. Alternatively, we can write a recursive function that calls itself with smaller numbers until it reaches the base cases (0 and 1).

```moonbit
// Loop form
let mut i = 0

> while i < 2, i = i + 1 {

Reviewer 2: mdlint run?

Author: already run. no warnings

Reviewer: @DawnMagnet Are you sure that you used the correct markdown linter (`@moonbit/markdown-linter`, not `mdlint`) to check the document?

# Modern Programming Concepts Overview: Imperative Programming

## Imperative vs. Functional Programming

- **Imperative**: Focuses on how to change program state through commands and variables.
- **Functional**: Views computation as the evaluation of expressions and avoids changing state.

## Core Concepts

- **Referential Transparency**: Functions and expressions can be replaced with their values without affecting program behavior.
- **Side Effects**: Actions that alter the program's state or external environment, which can include I/O operations.
- **Single-Value Types**: Functions that perform actions without returning a value, often used for side effects.

## Data Handling

- **Variables**: Stores data that can be mutable or immutable.
- **Structs**: Complex data structures with mutable fields, allowing for aliasing and direct manipulation of data.

## Control Structures

> - **Loops**: Repeatedly execute code based on a condition, with control over iteration and termination (break and continue).

Reviewer 3: ```suggestion
- **Loops**: Repeatedly execute code based on a condition, with control over iteration and termination (`break` and `continue`).
```

Author: already overwrite lec7

# Modern Programming Thoughts: Imperative Programming

## Revisiting Functional Programming

The core idea of functional programming is that each input corresponds to a fixed output, and identifiers can be directly substituted with their corresponding values, a property known as referential transparency. However, developing practical programs often requires "side effects" beyond calculations, such as input/output operations and modifying data in memory. These side effects may lead to inconsistent results upon multiple executions, breaking referential transparency.

```moonbit
let x: Int = 1 + 1 // x can be directly replaced with 2
fn square(x: Int) -> Int { x * x }
let z: Int = square(x) // can be replaced with { 2 * 2 }, still resulting in 4
```

## Commands and Side Effects

> The `print` command in Moonbite allows us to output strings, such as `print("hello moonbit")`. We can define initialization instructions in the `init` code block, serving as the program entry point. However, output commands may break referential transparency.

Reviewer 3: ```suggestion
The `print` command in MoonBit enables us to output strings, such as `print("hello moonbit")`. We can define initialization instructions in the `init` code block, serving as the program entry point. However, output commands may break referential transparency.
```

```moonbit
fn init {
  let x: Int = {
    println("hello moonbit") // First, execute the command and perform output
    1 + 1 // Then, take the last value of the expression block as the block's value
  }
  let z: Int = square(x) // Output once
}
```

We typically use the unit type `Unit` to represent the return value of functions or commands with side effects. It has only one value: `()`.

## Variables and Aliases

> In Moonbite, we can define temporary variables using `let mut`. The assignment operation itself is a command. Struct fields are immutable by default, but we can use `mut` to identify mutable fields, which are treated as references.

Reviewer 3: ```suggestion
In MoonBit, we can define temporary variables using `let mut`. The assignment operation itself is a command. Struct fields are immutable by default, but we can use `mut` to identify mutable fields, which are treated as references.
```

Reviewer 3: I think all "Moonbite" should be "MoonBit".

```moonbit
struct Ref[T] { mut val : T }

fn init {
  let ref: Ref[Int] = { val: 1 } // ref itself is just a data binding
  ref.val = 10 // We can modify the fields of the struct
  println(ref.val.to_string()) // Output 10
}
```

Multiple identifiers pointing to the same mutable data structure can be considered aliases, which need to be handled carefully.

## Loops

> We can define loops using variables in Moonbite. A loop includes defining the loop variable and its initial value, checking whether to continue the loop, and iterating the variable.

Reviewer 3: ```suggestion
We can define loops using variables in MoonBit. A loop includes defining the loop variable and its initial value, checking whether to continue the loop, and iterating the variable.
```

```moonbit
let mut i = 0
while i < 2, i = i + 1 {
  println("Output")
} // Repeat output 2 times
```

> The loop execution flow is: check the condition -> execute commands -> iterate the variable -> repeat the above process. Moonbite provides a debugger that allows observing the changes in runtime data during the loop process.

Reviewer 3: ```suggestion
The loop execution flow is: check the condition -> execute commands -> iterate the variable -> repeat the above process. MoonBit provides a debugger that allows observing the changes in runtime data during the loop process.
```

## Loops and Recursion

Loops and recursion are equivalent. A loop can be written in a recursive form. Taking the calculation of the Fibonacci sequence as an example:

```moonbit
// Loop form
fn fib_mut(n: Int) -> Int {
  let mut a = 0; let mut b = 1; let mut i = 0
  while i < n, i = i + 1 {
    let t = a + b
    a = b; b = t
  }
  a
}

// Recursive form
fn fib(n: Int) -> Int {
  if n == 0 { 0 }
  else if n == 1 { 1 }
  else { fib(n - 1) + fib(n - 2) }
}
```

## Controlling Loop Flow

Within a loop, we can use `break` to exit the loop prematurely and `continue` to skip the remaining part of the current loop iteration.

```moonbit
fn print_first_3() {
  let mut i = 0
  while i < 10, i = i + 1 {
    if i == 3 {
      break // Skip from 3 onwards
    } else {
      println(i.to_string())
    }
  }
}
```

> ## Moonbite Checks

Reviewer 3: ```suggestion
## MoonBit Checks
```

> Moonbite checks whether a variable is modified, which can help avoid mistakes like forgetting to add an iteration condition in a loop. It also checks if the function's return value type matches the declared type, preventing incorrect type declarations.

Reviewer 3: ```suggestion
MoonBit checks whether a variable is modified, which can help avoid mistakes like forgetting to add an iteration condition in a loop. It also checks if the function's return value type matches the declared type, preventing incorrect type declarations.
```

However, in the real world, we often need our programs to do more than just calculations. We might need them to read from a file, write to the screen, or change data in memory. These actions are called "side effects," and they can make our programs behave differently each time they run, even with the same input. This breaks the referential transparency, making our programs harder to understand and predict.

```moonbit
fn init {
  let x: Int = 1 + 1 // x can be directly replaced with 2
  fn square(x: Int) -> Int { x * x };
  let z: Int = square(x) // can be replaced with { 2 * 2 }, still resulting in 4

> }

Reviewer 3: Why not `moonbit expr`?

Reviewer: This is an init block so it can't be inside the `expr`

Having `expr` is equivalent to
```
fn init {
  <code goes here>
}
```

Reviewer 3: What I mean is that it seems to be better to mark the code block as `moonbit expr` instead of wrapping it with `fn init`. In the Chinese version, these codes are not wrapped by `fn init`.

Reviewer: Ok, I see, you are suggesting
```suggestion
```moonbit expr
let x: Int = 1 + 1 // x can be directly replaced with 2
fn square(x: Int) -> Int { x * x };
let z: Int = square(x) // can be replaced with { 2 * 2 }, still resulting in 4
```
Yeah, that looks better

Multiple identifiers pointing to the same mutable data structure can be considered aliases, which need to be handled carefully.

```moonbit
fn alter(a: Ref[Int], b: Ref[Int]) {
  a.val = 10
  b.val = 20
}

fn init {
  let x: Ref[Int] = { val : 1 }
  alter(x, x)
  println(x.val.to_string()) // x.val will be changed twice
}
```

## Debugger

- Moon Rabbit's debugger allows us to see real-time running data during operation and better understand the running process

![](../pics/debugger.png)

## Loops

We can define loops using variables in MoonBit. A loop includes defining the loop variable and its initial value, checking whether to continue the loop, and iterating the variable.

```moonbit
let mut i = 0
while i < 2 {
  println("Output")

>   i = i + 1

Reviewer 3: MoonBit recently added support for the `+=` operator. We can update our sample code to reflect this change.

```moonbit
// Loop form
fn fib_mut(n: Int) -> Int {
  let mut a = 0;
  let mut b = 1;
  let mut i = 0

>   while i < n, i = i + 1 {

Reviewer 3: The `while` loop here should also be updated.
